cdap-common/src/main/java/co/cask/cdap/common/utils/TimeMathParser.java

```diff
@@ -79,20 +82,57 @@ public static int resolutionInSeconds(String resolutionStr) {
     return output;
   }
 
+  /**
+   * Parses a time in String format into a long value. If the input is numeric we assume the input is in seconds.
+   *
+   * @param timeStr the string to parse
+   * @return the parsed time in seconds
+   */
   public static long parseTime(String timeStr) {
-    return parseTime(nowInSeconds(), timeStr);
+    return parseTime(timeStr, TimeUnit.SECONDS);
   }
 
+  /**
+   * Parses a time in String format into a long value
+   *
+   * @param timeStr the string to parse
+   * @param timeUnit the unit of time to return, if timeStr is numeric then it is assumed to be in the unit timeUnit
+   * @return the parsed time
+   */
+  public static long parseTime(String timeStr, TimeUnit timeUnit) {
+    return parseTime(nowInSeconds(), timeStr, timeUnit);
+  }
+
+  /**
+   * Parses a time in String format into a long value. If the input is numeric we assume the input is in seconds.
+   *
+   * @param now the present time in seconds
+   * @param timeStr the string to parse
+   * @return the parsed time in seconds
+   */
   public static long parseTime(long now, String timeStr) {
+    return parseTime(now, timeStr, TimeUnit.SECONDS);
+  }
+
+  /**
+   * Parses a time in String format into a long value
+   *
+   * @param now the present time in seconds
+   * @param timeStr the string to parse
+   * @param timeUnit the unit of time to return, if timeStr is numeric then it is assumed to be in the unit timeUnit
+   * @return the parsed time
+   * @throws IllegalArgumentException if the format of timeStr is bad
+   */
+  public static long parseTime(long now, String timeStr, TimeUnit timeUnit) {
```

Comment:
The reason this is necessary is that fetch requires the time be in milliseconds, where as parsing a string like now - 5s returns a time in seconds. Thus, we must either attempt to parse the long beforehand, (which terence was against), or move the logic into parseTime.